src/gasman.c

```diff
@@ -1099,6 +1099,27 @@ static void MaybeMoveBags(void)
 
 #endif
 
+/****************************************************************************
+**
+*F  OutOfMemoryPanic(size) . . . . . . . . . .  handle running out of memory
+**
+*/
+
+static TooLargeAllocationHandlerFunc TooLargeAllocFunc;
+void SetTooLargeAllocationHandler(TooLargeAllocationHandlerFunc f)
+{
+    TooLargeAllocFunc = f;
+}
+
+void OutOfMemoryPanic(UInt size, const char * reason)
+{
+    // If we could not satisfy an allocation, and that allocation was large
+    // (128MB at present), then GAP can reasonably continue.
+    if (size > (1LL << 27) && TooLargeAllocFunc) {
+        TooLargeAllocFunc(size, reason);
+    }
+    Panic("%s", reason);
```

Comment:
Note that Panic is a macro that is careful to record the site it was called at. It would be nice if `OutOfMemoryPanic` did that as well (it can then proceed to call `Panic_`). I.e., rename it to `OutOfMemoryPanic_` and then add a macro `OutOfMemoryPanic` similar to the macro `Panic`.

Then we can also do away with the "!!!" hack in the various panic messages.